clients/src/main/java/org/apache/kafka/common/utils/KafkaEventQueue.java

```diff
@@ -0,0 +1,390 @@
+                    if (entry != null) {
+                        long now = SystemTime.SYSTEM.nanoseconds();
+                        if (entry.getKey() <= now || closingTimeNs <= now) {
+                            timedEventContext = entry.getValue();
+                            remove(timedEventContext);
+                            continue;
+                        }
+                        awaitNs = entry.getKey() - now;
+                    }
+                    if (head.next == head) {
+                        if (closingTimeNs != Long.MAX_VALUE) {
+                            // If there are no more entries to process, and the queue is
+                            // closing, exit the thread.
+                            return;
+                        }
+                    } else {
+                        queuedEventContext = head.next;
+                        remove(queuedEventContext);
+                        continue;
+                    }
+                    if (closingTimeNs != Long.MAX_VALUE) {
+                        long now = SystemTime.SYSTEM.nanoseconds();
+                        if (awaitNs > closingTimeNs - now) {
+                            awaitNs = closingTimeNs - now;
+                        }
+                    }
+                    if (awaitNs == Long.MAX_VALUE) {
+                        cond.await();
+                    } else {
+                        cond.awaitNanos(awaitNs);
+                    }
+                } finally {
+                    lock.unlock();
+                }
+            }
+        }
+
+        void enqueue(EventContext<?> eventContext,
+                     Function<Long, Long> deadlineNsCalculator) {
+            lock.lock();
+            try {
+                Long existingDeadlineNs = null;
+                if (eventContext.tag != null) {
+                    EventContext<?> toRemove =
+                        tagToEventContext.put(eventContext.tag, eventContext);
+                    if (toRemove != null) {
+                        existingDeadlineNs = toRemove.deadlineNs;
+                        remove(toRemove);
+                        toRemove.cancel();
+                    }
+                }
+                Long deadlineNs = deadlineNsCalculator.apply(existingDeadlineNs);
+                boolean queueWasEmpty = head.isSingleton();
+                boolean shouldSignal = false;
+                switch (eventContext.insertionType) {
+                    case APPEND:
+                        head.insertBefore(eventContext);
+                        if (queueWasEmpty) {
+                            shouldSignal = true;
+                        }
+                        break;
+                    case PREPEND:
+                        head.insertAfter(eventContext);
+                        if (queueWasEmpty) {
+                            shouldSignal = true;
+                        }
```

Comment:
I think we signal regardless since we added this at the front of the queue?

Reply:
Actually, no, this is correct -- if the queue wasn't empty then the thread won't be waiting.